pyutilib/misc/timing.py

```diff
@@ -160,3 +160,316 @@ def start(self):
+                                                     time_per_call='Per Call(s)',
+                                                     relative_percent='% Time')
+            s += indent + '-' * (max_name_len + 60) + '\n'
+            other_time = self.total_time
+            for name, timer in self.timers.items():
+                s += indent
+                s += name_formatter.format(name=name)
+                s += '{0:>15.2e}'.format(timer.total_time)
+                s += '{0:>15d}'.format(timer.n_calls)
+                s += '{0:>15.2e}'.format(timer.total_time/timer.n_calls)
+                if self.total_time > 0:
+                    s += '{0:>15.1f}%\n'.format(timer.total_time / self.total_time * 100)
+                else:
+                    s += '{0:>15}\n'.format('nan')
+                s += timer.pprint(indent=indent + '    ')
+                other_time -= timer.total_time
+            s += indent
+            s += name_formatter.format(name='other')
+            s += '{0:>15.2e}'.format(other_time)
+            s += '{0:>15}'.format('N/A')
+            s += '{0:>15}'.format('N/A')
+            if self.total_time > 0:
+                s += '{0:>15.1f}%\n'.format(other_time / self.total_time * 100)
+            else:
+                s += '{0:>15}\n'.format('nan')
+        return s
+
+
+class HierarchicalTimer(object):
+    """
+    A class for hierarchical timing.
+
+    Examples
+    --------
+    >>> from pyutilib.misc.timing import HierarchicalTimer
+    >>> timer = HierarchicalTimer()
+    >>> timer.start('all')
+    >>> for i in range(10):
+    >>>     timer.start('a')
+    >>>     for i in range(5):
+    >>>         timer.start('aa')
+    >>>         timer.stop('aa')
+    >>>     timer.start('ab')
+    >>>     timer.stop('ab')
+    >>>     timer.stop('a')
+    >>> for i in range(10):
+    >>>     timer.start('b')
+    >>>     timer.stop('b')
+    >>> timer.stop('all')
+    >>> print(timer)
+    Identifier         Time (s)        # Calls   Per Call (s)
+    ---------------------------------------------------------
+    all                1.29e-01              1       1.29e-01
+        Identifier         Time (s)        # Calls    Per Call(s)         % Time
+        ------------------------------------------------------------------------
+        a                  1.15e-01             10       1.15e-02           89.0%
+            Identifier         Time (s)        # Calls    Per Call(s)         % Time
+            ------------------------------------------------------------------------
+            aa                 6.60e-02             50       1.32e-03           57.3%
+            ab                 4.84e-02             10       4.84e-03           42.0%
+            other              8.00e-04            N/A            N/A            0.7%
+        b                  1.39e-02             10       1.39e-03           10.8%
+        other              3.49e-04            N/A            N/A            0.3%
```

Comment:
Would you consider changing this format to more closely match the results from the Python profiler?  For reference, the profiler generates:
```
Function                                             called...
                                                         ncalls  tottime  cumtime
{posix.waitpid}                                      ->
sre_compile.py:228(_compile_charset)                 ->     122    0.026    0.065  sre_compile.py:256(_optimize_charset)
                                                            502    0.000    0.000  {method 'append' of 'list' objects}
                                                             20    0.000    0.000  {method 'extend' of 'list' objects}
```
I am thinking about a format similar to:
```
Identifier                               ncalls  tottime  cumtime      %
------------------------------------------------------------------------
all                                           1    0.129    0.129  100.0
    --------------------------------------------------------------------
    a                                        10    0.115    0.115   89.0
        ----------------------------------------------------------------
        aa                                   50    0.066    0.001   57.3
        ab                                   10    0.048    0.005   42.0
        other                               n/a    0.001      n/a    0.7
        ================================================================
    b                                        10    0.014    0.014   10.8
    other              3.49e-04       N/A            N/A             0.3
    ====================================================================
========================================================================
```
...I am happy to draft the patch if you want.

Reply:
I like this output way more than what I had. I honestly couldn't come up with anything I really liked, so I just gave up.

Reply:
I can take a stab at it right now.